app/models/katello/concerns/host_managed_extensions.rb

```diff
@@ -348,11 +348,21 @@ def import_enabled_repositories(repos)
       end
 
       def import_module_streams(module_streams)
+        # create_or_find_by avoids race conditions during concurrent registrations but clogs postgres logs with harmless errors.
+        # So we'll use create_or_find_by! during registration and first_or_create! otherwise.
+        registered_time = subscription_facet&.registered_at
+        use_create_or_find_by = registered_time.nil? || registered_time > 1.minute.ago
         streams = {}
         module_streams.each do |module_stream|
```

Comment:
I know this code is old, but I wonder if now would be a good time to switch to `insert_all` or `upsert_all`. It feels odd to do so many creations one postgres insert at a time in a Ruby loop.

Bulk insertion lets you create all entities in a single Postgres insert call. It takes into account duplicates too from what I remember. You can pass in what the records should be `unique_by` (in this case name, stream, and context).

There might be a catch, but I think it's worth a thought.

https://apidock.com/rails/v6.0.0/ActiveRecord/Persistence/ClassMethods/insert_all
https://apidock.com/rails/v6.0.0/ActiveRecord/Persistence/ClassMethods/upsert_all

We use them in a couple places: 

https://github.com/search?q=repo%3AKatello%2Fkatello%20insert_all&type=code
https://github.com/search?q=repo%3AKatello%2Fkatello+upsert_all&type=code